pdm/models/candidates.py

```diff
@@ -215,7 +215,8 @@ def get_metadata(
         return self.metadata
 
     def __repr__(self) -> str:
-        return f"<Candidate {self.name} {self.version}>"
+        source = getattr(self.link, "comes_from", "unknown")
```

Comment:
What is the value if the candidate is from a file link(including VCS, remote file and local file)?

Reply:
Good question. I haven't tried and don't know. At least it doesn't break if `comes_from` is not set.

Reply:
Just to make sure it displays a meaningful value not confusing

Reply:
For the cases you mentioned, pdm doesn't seem to go through the repository search. At least I wasn't able to get the candidate list output.